File: OnlyCue/LTC/LTCFrameStream.swift
import Foundation

/// A continuous LTC audio stream — successive 80-bit SMPTE frames starting at
/// `startTimecode`, biphase-mark polarity threaded across the frame boundaries
/// so the concatenated output is a seamless waveform. Pure value type; the
/// `AVAudioEngine` that schedules these samples onto an output device is a
/// separate concern (a later leaf of epic #33).
struct LTCFrameStream {

    let startTimecode: Timecode
    let sampleRate: Double
    var amplitude: Float

    init(
        startTimecode: Timecode,
        sampleRate: Double,
        amplitude: Float = LTCEncoder.defaultAmplitude
    ) {
        precondition(sampleRate > 0, "sample rate must be positive")
        self.startTimecode = startTimecode
        self.sampleRate = sampleRate
        self.amplitude = amplitude
    }

    var framesPerSecond: Int { startTimecode.rate.framesPerSecond }

    /// Samples in one LTC frame at this stream's sample rate (constant across
    /// frames — `round(sampleRate / fps)`).
    var samplesPerFrame: Int { Int((sampleRate / Double(framesPerSecond)).rounded()) }

    /// The timecode of the frame `offset` frames after the start (negative
    /// offsets clamp to the start).
    func timecode(atFrameOffset offset: Int) -> Timecode {
        Timecode(frameCount: startTimecode.frameCount + max(0, offset), rate: startTimecode.rate)
    }

    /// Float PCM (mono, `±amplitude`) for `count` consecutive frames beginning
    /// at `startTimecode`, with biphase polarity carried across the joins so the
    /// result has no boundary glitch. Empty when `count <= 0`.
    func samples(frameCount count: Int) -> [Float] {
        guard count > 0 else { return [] }
        var output: [Float] = []
        output.reserveCapacity(count * samplesPerFrame + count)
        var level = false
        for offset in 0..<count {
            let (frameSamples, endLevel) = LTCEncoder.samples(
                for: timecode(atFrameOffset: offset),
                sampleRate: sampleRate,
                amplitude: amplitude,
                startLevel: level
            )
            output.append(contentsOf: frameSamples)
            level = endLevel
        }
        return output
    }
}

File: windows/OnlyCue.Core/Ltc/LtcFrameStream.cs
namespace OnlyCue.Core.Ltc;

/// <summary>
/// A continuous LTC audio stream — successive 80-bit SMPTE frames starting at
/// <see cref="StartTimecode"/>, biphase-mark polarity threaded across the frame
/// boundaries so the concatenated output is a seamless waveform.
/// </summary>
/// <remarks>
/// <para>
/// Mirrors the Swift <c>LTCFrameStream</c> (<c>OnlyCue/LTC/LTCFrameStream.swift</c>);
/// macOS is the source of truth and <c>golden/ltc-schedule-v1.json</c> is the
/// contract.
/// </para>
/// <para>
/// Each frame is encoded from the level the previous one ended at, so no
/// transition is lost at a join. In practice the threading never changes the
/// output — every LTC frame ends at the level it started at, because parity
/// forces an even flip count — so this is a safety net rather than a live
/// correction, and the contract's <c>joins</c> windows cannot catch its removal.
/// What they do catch is a join whose polarity is <i>inverted</i>: that loses a
/// transition, which is a misread bit rather than a harmless inversion.
/// </para>
/// </remarks>
public sealed class LtcFrameStream
{
    /// <summary>Swift traps on a non-positive sample rate (<c>precondition</c>);
    /// throwing keeps a bad caller failing on both sides.</summary>
    public LtcFrameStream(Timecode startTimecode, double sampleRate, float amplitude = LtcEncoder.DefaultAmplitude)
    {
        if (!(sampleRate > 0))
        {
            throw new ArgumentOutOfRangeException(nameof(sampleRate), sampleRate, "sample rate must be positive");
        }

        StartTimecode = startTimecode;
        SampleRate = sampleRate;
        Amplitude = amplitude;
    }

    public Timecode StartTimecode { get; }
    public double SampleRate { get; }
    public float Amplitude { get; }

    public int FramesPerSecond => StartTimecode.Rate.FramesPerSecond();

    /// <summary>Samples in one LTC frame at this stream's sample rate (constant
    /// across frames — <c>round(sampleRate / fps)</c>, half away from zero).</summary>
    public int SamplesPerFrame => (int)Math.Round(SampleRate / FramesPerSecond, MidpointRounding.AwayFromZero);

    /// <summary>The timecode of the frame <paramref name="offset"/> frames after
    /// the start (negative offsets clamp to the start).</summary>
    public Timecode TimecodeAtFrameOffset(int offset) =>
        Timecode.FromFrameCount(StartTimecode.FrameCount + Math.Max(0, offset), StartTimecode.Rate);

    /// <summary>Float PCM (mono, <c>±amplitude</c>) for <paramref name="count"/>
    /// consecutive frames beginning at <see cref="StartTimecode"/>, with biphase
    /// polarity carried across the joins. Empty when <paramref name="count"/> is
    /// not positive.</summary>
    public float[] Samples(int count)
    {
        if (count <= 0)
        {
            return [];
        }

        var output = new List<float>(count * SamplesPerFrame + count);
        var level = false;
        for (var offset = 0; offset < count; offset++)
        {
            var (frameSamples, endLevel) = LtcEncoder.Samples(
                TimecodeAtFrameOffset(offset),
                SampleRate,
                Amplitude,
                level);
            output.AddRange(frameSamples);
            level = endLevel;
        }

        return output.ToArray();
    }
}
